File: frontend/src/pages/imports/firefly/utils/rowResolution.ts
import type { AccountsOverview } from '@/api/accounts'
import type { Category } from '@/api/categories'
import { FIREFLY_NO_CATEGORY_SOURCE } from '@/api/firefly-imports'
import type { Institution } from '@/api/institutions'
import { CREATE_ACCOUNT_VALUE, CREATE_CATEGORY_VALUE, DEFAULT_CATEGORY_ICON } from '@/pages/imports/constants'
import type { Currency } from '@/api/currency'
import type { CsvRow, ImportCategoryKind } from '@/pages/imports/types'
import { MAX_IMPORT_MINOR_UNITS, toImportMinorUnits } from '@/pages/imports/utils'
import { findReusedImportCategory } from '@/pages/imports/utils/categoryMatching'
import { findCurrencyExponent } from '@/utils/moneyInput'
import {
  FIREFLY_BALANCE_ROW_UNATTACHED_REASON,
  FIREFLY_DEPOSIT_DESTINATION_UNTRACKED_REASON,
  FIREFLY_GENERIC_SKIP_REASON,
  FIREFLY_TRANSFER_ENDPOINT_UNTRACKED_REASON,
  FIREFLY_TYPE_DEPOSIT,
  FIREFLY_TYPE_OPENING_BALANCE,
  FIREFLY_TYPE_RECONCILIATION,
  FIREFLY_TYPE_WITHDRAWAL,
  FIREFLY_WITHDRAWAL_SOURCE_UNTRACKED_REASON,
  getFireflyUnsupportedTypeReason,
  isFireflyJournalType,
} from '@/pages/imports/firefly/constants'
import type { FireflyAccountSource, FireflyAccountSources } from '@/pages/imports/firefly/types'
import { getFireflyRowAmounts } from './derivation'
import type { FireflyAccountCreateDetails } from './payload'

/**
 * Mapping lookups needed to resolve journal rows the same way the commit will
 */
export interface FireflyRowResolutionOptions {
  /** The accounts rows are written to, which the account mappings and create details are keyed by */
  accountSources: FireflyAccountSources
  accountById: Map<string, AccountsOverview>
  accountMappings: Record<string, string>
  accountCreateDetails: Record<string, FireflyAccountCreateDetails>
  institutionById: Map<string, Institution>
  categoryById: Map<string, Category>
  categoryMappings: Record<string, string>
  categoryCreateKinds: Record<string, ImportCategoryKind>
  transferCategory: Category | undefined
  balanceAdjustmentCategory: Category | undefined

  /** Currency metadata, read for the decimal places each account's currency holds */
  currencies: Currency[]
}

/**
 * Ledger account details one tracked journal endpoint resolves to after the
 * user's mapping choices are applied
 */
export interface FireflyResolvedAccount {
  id: string
  name: string
  currency: string
  institution: Institution | null
}

/**
 * One ledger transaction a journal row produces
 */
export interface FireflyResolvedLeg {
  account: FireflyResolvedAccount
  amount: number
  category: Category | undefined
  merchantName: string | null

  /** Account the money moved to or from, held only by the two legs of a transfer */
  counterpartyAccount: FireflyResolvedAccount | null
}

/**
 * Outcome of resolving one journal row: the ledger legs the import will create, the reason the
 * import will skip the row, or neither while an account the row writes to has no answer yet
 */
export type FireflyRowResolution =
  | { legs: FireflyResolvedLeg[]; skipReason: null }
  | { legs: null; skipReason: string }
  | { legs: null; skipReason: null }

/**
 * Internal signal that a row cannot be converted, mirroring the backend's
 * skip error so both sides walk the same decision order
 */
class FireflyRowSkipError extends Error {
  readonly reason: string

  constructor(reason: string) {
    super(reason)
    this.reason = reason
  }
}

/**
 * Internal signal that a row writes to an imported account the user has not answered yet
 */
class FireflyAccountUnansweredError extends Error {}

/**
 * Resolves one journal row into the ledger legs the import will create, or
 * the backend-worded reason the import will skip the row
 */
export function resolveFireflyRowLegs(row: CsvRow, options: FireflyRowResolutionOptions): FireflyRowResolution {
  try {
    return { legs: buildFireflyRowLegs(row, options), skipReason: null }
  } catch (error) {
    if (error instanceof FireflyRowSkipError) return { legs: null, skipReason: error.reason }

    // The import stays blocked until the account is answered, and the answer decides the outcome
    if (error instanceof FireflyAccountUnansweredError) return { legs: null, skipReason: null }

    // A row failing in a way no skip rule anticipated must not break the
    // preview, so it is predicted as skipped with the same generic reason
    // the backend reports when its own fallback catches the row
    return { legs: null, skipReason: FIREFLY_GENERIC_SKIP_REASON }
  }
}

/**
 * Returns the mapped category a successful resolution will commit, or undefined when the row's
 * resolver used Transfer or Balance Adjustment instead
 */
export function getFireflyCategoryUsedByResolution(
  row: CsvRow,
  legs: FireflyResolvedLeg[],
  options: FireflyRowResolutionOptions,
) {
  const mappedCategory = getFireflyMappedCategory(row, options)
  if (!mappedCategory || !legs.some((leg) => leg.category?.id === mappedCategory.id)) return undefined

  const source = row.category?.trim() || FIREFLY_NO_CATEGORY_SOURCE
  if (options.categoryMappings[source] !== CREATE_CATEGORY_VALUE) return mappedCategory

  const reused = findReusedImportCategory(source, options.categoryById.values())
  return reused?.kind === mappedCategory.kind ? reused : undefined
}

/**
 * Builds the ledger legs for one journal row, throwing a skip error in the
 * same branch order as the backend resolver so predicted reasons match
 */
function buildFireflyRowLegs(row: CsvRow, options: FireflyRowResolutionOptions): FireflyResolvedLeg[] {
  const journalType = row.type?.trim().toLowerCase() ?? ''

  // A type the importer does not know is refused before any account is resolved, since the
  // transfer rule below would otherwise write it as a transfer between two imported accounts
  if (!isFireflyJournalType(journalType)) {
    throw new FireflyRowSkipError(getFireflyUnsupportedTypeReason(row.type?.trim() ?? ''))
  }

  const sourceAccountSource = options.accountSources.find(row.source_name, row.source_type)
  const destinationAccountSource = options.accountSources.find(row.destination_name, row.destination_type)
  const source = resolveFireflyMappedAccount(sourceAccountSource, options)
  const destination = resolveFireflyMappedAccount(destinationAccountSource, options)

  // Firefly III pairs balance rows with a virtual balance account, so the
  // imported side is whichever endpoint is a real account and money flowing
  // into it is positive
  if (journalType === FIREFLY_TYPE_OPENING_BALANCE || journalType === FIREFLY_TYPE_RECONCILIATION) {
    const account = destination ?? source
    if (!account) throw new FireflyRowSkipError(FIREFLY_BALANCE_ROW_UNATTACHED_REASON)

    const amount = getFireflyAmountInAccountCurrency(row, account.currency, options.currencies)
    return [{
      account,
      amount: destination ? amount : -amount,
      category: options.balanceAdjustmentCategory,
      merchantName: null,
      counterpartyAccount: null,
    }]
  }

  // A journal between two imported accounts is a transfer in Lumina no matter
  // the Firefly III type, which covers loan payments recorded as withdrawals
  // into a liability account
  if (source && destination) {
    // Two accounts in the file can be mapped onto one account, which is how a
    // renamed account is carried across. The pair would then be two cancelling
    // rows in that account, a shape the API refuses when entered by hand.
    // Two different accounts queued for creation share the create sentinel as
    // their id and still become two separate accounts, so their sources decide it
    const isSameCreate = sourceAccountSource?.id === destinationAccountSource?.id
    if (source.id === destination.id && (source.id !== CREATE_ACCOUNT_VALUE || isSameCreate)) {
      throw new FireflyRowSkipError('Transfer source and destination resolve to the same account')
    }

    return [
      {
        account: source,
        amount: -getFireflyAmountInAccountCurrency(row, source.currency, options.currencies),
        category: options.transferCategory,
        merchantName: null,
        counterpartyAccount: destination,
      },
      {
        account: destination,
        amount: getFireflyAmountInAccountCurrency(row, destination.currency, options.currencies),
        category: options.transferCategory,
        merchantName: null,
        counterpartyAccount: source,
      },
    ]
  }

  if (journalType === FIREFLY_TYPE_WITHDRAWAL) {
    if (!source) throw new FireflyRowSkipError(FIREFLY_WITHDRAWAL_SOURCE_UNTRACKED_REASON)

    return [{
      account: source,
      amount: -getFireflyAmountInAccountCurrency(row, source.currency, options.currencies),
      category: getFireflyMappedCategory(row, options),
      merchantName: row.destination_name?.trim() || null,
      counterpartyAccount: null,
    }]
  }

  if (journalType === FIREFLY_TYPE_DEPOSIT) {
    if (!destination) throw new FireflyRowSkipError(FIREFLY_DEPOSIT_DESTINATION_UNTRACKED_REASON)

    return [{
      account: destination,
      amount: getFireflyAmountInAccountCurrency(row, destination.currency, options.currencies),
      category: getFireflyMappedCategory(row, options),
      merchantName: row.source_name?.trim() || null,
      counterpartyAccount: null,
    }]
  }

  throw new FireflyRowSkipError(FIREFLY_TRANSFER_ENDPOINT_UNTRACKED_REASON)
}

/**
 * Resolves a tracked journal endpoint to the ledger account the mapping
 * choices produce, or null when the endpoint is not an imported account.
 * Throws while the endpoint is an imported account with no answer yet
 */
function resolveFireflyMappedAccount(
  accountSource: FireflyAccountSource | null,
  options: FireflyRowResolutionOptions,
): FireflyResolvedAccount | null {
  if (!accountSource) return null

  const choice = options.accountMappings[accountSource.id]
  if (!choice) throw new FireflyAccountUnansweredError()
  if (choice === CREATE_ACCOUNT_VALUE) {
    const details = options.accountCreateDetails[accountSource.id]
    return {
      id: CREATE_ACCOUNT_VALUE,
      name: accountSource.name,
      currency: (details?.currency ?? '').trim().toUpperCase(),
      institution: options.institutionById.get(details?.institutionId ?? '') ?? null,
    }
  }

  const account = options.accountById.get(choice)
  if (!account) return null
  return { id: account.id, name: account.name, currency: account.currency, institution: account.institution }
}

/**
 * Gets the row's absolute amount in account-currency minor units, throwing
 * the backend-worded skip error when no amount exists in that currency, the
 * matching raw amount is unparseable, or its magnitude cannot be stored
 */
function getFireflyAmountInAccountCurrency(row: CsvRow, accountCurrency: string, currencies: Currency[]): number {
  // Firefly III writes the journal amount in the transaction currency and
  // carries a foreign amount when a second currency is involved, so the
  // account-side value is whichever of the two matches the account currency.
  // Both are read as the payload sends them
  const { main, foreign } = getFireflyRowAmounts(row)

  let rawAmount: string
  if (accountCurrency && main?.currencyCode === accountCurrency) {
    rawAmount = main.amount
  } else if (foreign && foreign.currencyCode === accountCurrency) {
    rawAmount = foreign.amount
  } else {
    throw new FireflyRowSkipError(`Neither the amount nor the foreign amount is in the account's currency (${accountCurrency})`)
  }

  // Every account's currency is one the API served, so a missing exponent is unreachable. An
  // unanticipated failure is reported with the generic reason rather than a skip rule of its own
  const exponent = findCurrencyExponent(currencies, accountCurrency)
  if (exponent === null) throw new Error(`No currency metadata for ${accountCurrency}`)

  // Explain excess precision separately while malformed text and parser-range overflow retain the raw value
  const minorUnits = toImportMinorUnits(rawAmount, exponent)
  if (minorUnits === 'tooPrecise') {
    throw new FireflyRowSkipError(
      `The amount has more decimal places than ${accountCurrency} has. A period is read as a decimal point, never as a separator between thousands.`,
    )
  }
  if (minorUnits === 'unreadable' || minorUnits === 'tooLarge') {
    throw new FireflyRowSkipError(`Invalid amount "${rawAmount}"`)
  }

  // The import writes the magnitude rather than the parsed value, and the signed range holds one
  // more value below zero than above it, so the smallest amount the parser accepts negates into
  // one the column cannot take. Such a row is left out, since its unsigned amount would be past
  // what the endpoint stores
  const absoluteMinorUnits = minorUnits < 0n ? -minorUnits : minorUnits
  if (absoluteMinorUnits > MAX_IMPORT_MINOR_UNITS) throw new FireflyRowSkipError(`Amount is too large: "${rawAmount}"`)

  return Number(absoluteMinorUnits)
}

/**
 * Gets the category the user's mapping choices assign to a categorized row
 */
function getFireflyMappedCategory(
  row: CsvRow,
  options: FireflyRowResolutionOptions,
): Category | undefined {
  const source = row.category?.trim() || FIREFLY_NO_CATEGORY_SOURCE
  const choice = options.categoryMappings[source]

  if (choice === CREATE_CATEGORY_VALUE) {
    return {
      id: `firefly-preview-category-${source}`,
      group_id: null,
      owner_id: null,
      name: source,
      kind: options.categoryCreateKinds[source] ?? 'expense',
      icon: DEFAULT_CATEGORY_ICON,
      is_system: false,
      created_at: '',
    }
  }

  return choice ? options.categoryById.get(choice) : undefined
}

File: backend/app/services/transactions/imports/firefly/row_resolution.py
"""Firefly III journal row to Lumina transaction leg resolution"""

import uuid
from dataclasses import dataclass
from datetime import date

from app.models.account import Account
from app.models.category import Category
from app.models.currency import Currency
from app.schemas.data_imports import FireflyTransactionRow
from app.services.transactions.imports.firefly.constants import (
    FIREFLY_NO_CATEGORY_SOURCE,
    FIREFLY_TRACKED_ACCOUNT_TYPES,
    FIREFLY_TYPE_DEPOSIT,
    FIREFLY_TYPE_OPENING_BALANCE,
    FIREFLY_TYPE_RECONCILIATION,
    FIREFLY_TYPE_TRANSFER,
    FIREFLY_TYPE_WITHDRAWAL,
)
from app.services.transactions.imports.shared.row_mappings import (
    get_import_row_account,
    get_import_row_category,
    validate_import_category_can_be_used_for_account,
)
from app.utils.money import (
    DecimalAmountParseError,
    DecimalAmountPrecisionError,
    parse_decimal_amount_to_minor_units,
)


class FireflyRowSkipError(Exception):
    """Raised when a Firefly III row cannot be converted into Lumina legs"""

    def __init__(self, reason: str) -> None:
        """Store the client-facing skip reason

        Args:
            reason: Why the row cannot be converted
        """
        super().__init__(reason)
        self.reason = reason


@dataclass
class FireflyResolutionContext:
    """Store lookups needed to resolve Firefly III rows into transaction legs

    Attributes:
        user_id: Identifier for the user running the import
        accounts_by_source: Account rows keyed by Firefly III account name
        categories_by_source: Category rows keyed by Firefly III category name
        currencies_by_code: Currency rows keyed by currency code
        transfer_category: System category applied to two-leg transfers
        balance_adjustment_category: System category applied to opening balances
    """

    user_id: uuid.UUID
    accounts_by_source: dict[str, Account]
    categories_by_source: dict[str, Category]
    currencies_by_code: dict[str, Currency]
    transfer_category: Category
    balance_adjustment_category: Category


@dataclass
class FireflyLeg:
    """One Lumina transaction produced from a Firefly III journal row

    Attributes:
        account: Account the transaction is written to
        dt: Transaction date
        amount: Signed amount in account-currency minor units
        category: Category applied to the transaction
        merchant_name: Optional counterparty recorded as a merchant
        notes: Optional combined description and notes text
        tag_names: Tag names applied to the transaction
    """

    account: Account
    dt: date
    amount: int
    category: Category
    merchant_name: str | None
    notes: str | None
    tag_names: list[str]


def resolve_firefly_row(row: FireflyTransactionRow, context: FireflyResolutionContext) -> list[FireflyLeg]:
    """Resolve one Firefly III journal row into Lumina transaction legs

    Args:
        row: Firefly III journal row from the import payload
        context: Lookups needed to resolve the row

    Returns:
        Transaction legs the row produces

    Raises:
        FireflyRowSkipError: Raised when the row cannot be converted
        HTTPException: Raised with 422 when a tracked account or category is not mapped
    """
    journal_type = row.type.strip().lower()
    source_account = _get_tracked_account(row.source_name, row.source_type, context)
    destination_account = _get_tracked_account(row.destination_name, row.destination_type, context)
    notes = _build_leg_notes(row)

    if journal_type in (FIREFLY_TYPE_OPENING_BALANCE, FIREFLY_TYPE_RECONCILIATION):
        return _resolve_balance_row(row, source_account, destination_account, notes, context)

    # A journal between two imported accounts is a transfer in Lumina no
    # matter the Firefly type, which covers loan payments recorded as
    # withdrawals into a liability account
    if source_account is not None and destination_account is not None:
        return _resolve_transfer_pair(row, source_account, destination_account, notes, context)

    if journal_type == FIREFLY_TYPE_WITHDRAWAL:
        if source_account is None:
            raise FireflyRowSkipError("Withdrawal source is not an imported account")
        category = _resolve_row_category(row, source_account, context)
        return [FireflyLeg(
            account=source_account,
            dt=row.dt,
            amount=-_get_amount_in_account_currency(row, source_account, context),
            category=category,
            merchant_name=_clean_name(row.destination_name),
            notes=notes,
            tag_names=row.tag_names,
        )]

    if journal_type == FIREFLY_TYPE_DEPOSIT:
        if destination_account is None:
            raise FireflyRowSkipError("Deposit destination is not an imported account")
        category = _resolve_row_category(row, destination_account, context)
        return [FireflyLeg(
            account=destination_account,
            dt=row.dt,
            amount=_get_amount_in_account_currency(row, destination_account, context),
            category=category,
            merchant_name=_clean_name(row.source_name),
            notes=notes,
            tag_names=row.tag_names,
        )]

    if journal_type == FIREFLY_TYPE_TRANSFER:
        raise FireflyRowSkipError("Transfer endpoint is not an imported account")

    raise FireflyRowSkipError(
        f'Journal type "{row.type.strip()}" is not supported, the importer handles'
        " withdrawals, deposits, transfers, opening balances, and reconciliations",
    )


def _resolve_transfer_pair(
    row: FireflyTransactionRow,
    source_account: Account,
    destination_account: Account,
    notes: str | None,
    context: FireflyResolutionContext,
) -> list[FireflyLeg]:
    """Resolve a row between two imported accounts into transfer legs

    Args:
        row: Firefly III journal row from the import payload
        source_account: Imported account money leaves
        destination_account: Imported account money enters
        notes: Combined description and notes text
        context: Lookups needed to resolve the row

    Returns:
        Outgoing and incoming transfer legs
    """
    return [
        FireflyLeg(
            account=source_account,
            dt=row.dt,
            amount=-_get_amount_in_account_currency(row, source_account, context),
            category=context.transfer_category,
            merchant_name=None,
            notes=notes,
            tag_names=row.tag_names,
        ),
        FireflyLeg(
            account=destination_account,
            dt=row.dt,
            amount=_get_amount_in_account_currency(row, destination_account, context),
            category=context.transfer_category,
            merchant_name=None,
            notes=notes,
            tag_names=row.tag_names,
        ),
    ]


def _resolve_balance_row(
    row: FireflyTransactionRow,
    source_account: Account | None,
    destination_account: Account | None,
    notes: str | None,
    context: FireflyResolutionContext,
) -> list[FireflyLeg]:
    """Resolve an opening balance or reconciliation row into one adjustment leg

    Firefly III pairs these rows with a virtual initial balance or
    reconciliation account, so the imported side is whichever endpoint is a
    real account. Money flowing into the imported side is positive

    Args:
        row: Firefly III journal row from the import payload
        source_account: Imported account on the source side when present
        destination_account: Imported account on the destination side when present
        notes: Combined description and notes text
        context: Lookups needed to resolve the row

    Returns:
        Single balance adjustment leg

    Raises:
        FireflyRowSkipError: Raised when neither endpoint is an imported account
    """
    account = destination_account or source_account
    if account is None:
        raise FireflyRowSkipError("Opening balance or reconciliation row is not attached to an imported account")

    amount = _get_amount_in_account_currency(row, account, context)
    return [FireflyLeg(
        account=account,
        dt=row.dt,
        amount=amount if destination_account is not None else -amount,
        category=context.balance_adjustment_category,
        merchant_name=None,
        notes=notes,
        tag_names=row.tag_names,
    )]


def _get_tracked_account(
    name: str | None,
    account_type: str | None,
    context: FireflyResolutionContext,
) -> Account | None:
    """Return the mapped account for a tracked journal endpoint

    Args:
        name: Account name on the journal endpoint
        account_type: Firefly III account type on the journal endpoint
        context: Lookups needed to resolve the row

    Returns:
        Mapped account when the endpoint is a tracked type, otherwise None

    Raises:
        HTTPException: Raised with 422 when a tracked account name is not mapped
    """
    if not name or not account_type:
        return None
    if account_type.strip().lower() not in FIREFLY_TRACKED_ACCOUNT_TYPES:
        return None
    return get_import_row_account(context.accounts_by_source, name)


def _resolve_row_category(
    row: FireflyTransactionRow,
    account: Account,
    context: FireflyResolutionContext,
) -> Category:
    """Return the mapped category for a categorized row

    Args:
        row: Firefly III journal row from the import payload
        account: Account the resulting transaction is written to
        context: Lookups needed to resolve the row

    Returns:
        Category mapped to the row's category name or the no-category source

    Raises:
        HTTPException: Raised with 422 when the category is not mapped or not usable
    """
    category_name = (row.category or "").strip() or FIREFLY_NO_CATEGORY_SOURCE
    category = get_import_row_category(context.categories_by_source, category_name)
    validate_import_category_can_be_used_for_account(category, account, context.user_id)
    return category


def _get_amount_in_account_currency(
    row: FireflyTransactionRow,
    account: Account,
    context: FireflyResolutionContext,
) -> int:
    """Return the row's absolute amount in the account's currency minor units

    Firefly III writes journal amounts in the transaction currency and carries
    a foreign amount when a second currency is involved, so the account-side
    value is whichever of the two matches the account currency

    Args:
        row: Firefly III journal row from the import payload
        account: Imported account one leg is written to
        context: Lookups needed to resolve the row

    Returns:
        Absolute amount in account-currency minor units

    Raises:
        FireflyRowSkipError: Raised when no amount is available in the account currency
    """
    if row.currency_code.upper() == account.currency:
        raw_amount = row.amount
    elif row.foreign_currency_code and row.foreign_amount and row.foreign_currency_code.upper() == account.currency:
        raw_amount = row.foreign_amount
    else:
        raise FireflyRowSkipError(
            f"Neither the amount nor the foreign amount is in the account's currency ({account.currency})",
        )

    currency = context.currencies_by_code[account.currency]
    try:
        amount = parse_decimal_amount_to_minor_units(
            raw_amount,
            currency_code=currency.id,
            minor_unit_exponent=currency.minor_unit_exponent,
        )
    except (DecimalAmountParseError, DecimalAmountPrecisionError) as exc:
        raise FireflyRowSkipError(f'Invalid amount "{raw_amount}"') from exc
    return abs(amount)


def _build_leg_notes(row: FireflyTransactionRow) -> str | None:
    """Return combined description and notes text for a row's legs

    Args:
        row: Firefly III journal row from the import payload

    Returns:
        Description and notes joined on separate lines, or None when both are empty
    """
    parts = [part.strip() for part in (row.description, row.notes) if part and part.strip()]
    return "\n".join(parts) or None


def _clean_name(name: str | None) -> str | None:
    """Return a trimmed counterparty name or None when empty

    Args:
        name: Raw counterparty name from a journal endpoint

    Returns:
        Trimmed name or None
    """
    cleaned = (name or "").strip()
    return cleaned or None
